File: app/src/am_hoanghh/issues8/UserAdapter.java
package issues8;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.asian.R;
import com.example.asian.databinding.DialogConfirmDeleteBinding;
import com.example.asian.databinding.ItemRvUserBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class UserAdapter extends ListAdapter<UserModel, UserAdapter.ViewHolder> {
    private final DBHandler dbHandler;
    private final Context context;
    private static final String DEBUG = "DEBUG";

    protected UserAdapter(DBHandler dbHandler, Context context) {
        super(DIFF_CALLBACK);
        this.dbHandler = dbHandler;
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemRvUserBinding binding = ItemRvUserBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new UserAdapter.ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        UserModel userModel = getItem(position);

        holder.binding.tvCountNumber.setText(String.valueOf(userModel.getCountNumber()));
        holder.binding.tvUsername.setText(userModel.getName());
        holder.binding.tvAge.setText(context.getString(R.string.textview_text_years_old, userModel.getAge()));

        Dialog dialog = new Dialog(context);

        holder.binding.ivDeleteUser.setOnClickListener(v -> {
            Log.d(DEBUG, String.valueOf(userModel.getId()));

            Objects.requireNonNull(dialog.getWindow()).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            DialogConfirmDeleteBinding dialogBinding = DialogConfirmDeleteBinding.inflate(LayoutInflater.from(context));
            dialog.setContentView(dialogBinding.getRoot());
            Objects.requireNonNull(dialog.getWindow()).setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            dialog.setCancelable(false);

            dialogBinding.tvExplain.setText(context.getString(R.string.textview_text_are_you_sure_you_want_to_delete, userModel.getName()));
            dialogBinding.tvCancel.setOnClickListener(view -> {
                dialog.dismiss();
            });
            dialogBinding.tvConfirm.setOnClickListener(view -> {
                dialog.dismiss();
                dbHandler.deleteUserById(userModel.getId());
                List<UserModel> oldLists = new ArrayList<>(getCurrentList());
                oldLists.remove(holder.getAdapterPosition());

                List<UserModel> newLists = new ArrayList<>();
                for (int i = 0; i < oldLists.size(); i++) {
                    UserModel item = oldLists.get(i);
                    newLists.add(new UserModel(item.getId(), item.getName(), item.getAge(), i + 1));
                }
                submitList(newLists);
            });
            dialog.show();
        });
    }

    public static final DiffUtil.ItemCallback<UserModel> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<UserModel>() {
                @Override
                public boolean areItemsTheSame(@NonNull UserModel oldItem, @NonNull UserModel newItem) {
                    return Objects.equals(oldItem.getId(), newItem.getId());
                }

                @Override
                public boolean areContentsTheSame(@NonNull UserModel oldItem, @NonNull UserModel newItem) {
                    return Objects.equals(oldItem.getName(), newItem.getName())
                            && Objects.equals(oldItem.getAge(), newItem.getAge())
                            && Objects.equals(oldItem.getCountNumber(), newItem.getCountNumber());
                }
            };

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemRvUserBinding binding;

        public ViewHolder(ItemRvUserBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/am_hoanghh/issues13/ex1/viewmodel/UserAdapter.kt
package issues13.ex1.viewmodel

import android.app.AlertDialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.asian.R
import com.example.asian.databinding.DialogConfirmDeleteBinding
import com.example.asian.databinding.ItemRvUserBinding
import issues13.ex1.model.UserModel

class UserAdapter(private val onDeleteListener: OnDeleteListener) :
    ListAdapter<UserModel, UserAdapter.ViewHolder>(DIFF_CALLBACK) {

    inner class ViewHolder(
        private val binding: ItemRvUserBinding
    ) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(user: UserModel) {
            val context: Context = itemView.context

            binding.tvCountNumber.text = user.countNumber.toString()
            binding.tvUsername.text = user.name
            binding.tvAge.text = context.getString(R.string.textview_text_years_old, user.age)

            val dialogBinding: DialogConfirmDeleteBinding = DialogConfirmDeleteBinding.inflate(
                LayoutInflater.from(context)
            )
            val dialog = AlertDialog.Builder(context).apply {
                setView(dialogBinding.root)
                setCancelable(false)
            }.create()
            dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            dialogBinding.tvCancel.setOnClickListener {
                dialog.dismiss()
            }
            dialogBinding.tvConfirm.setOnClickListener {
                dialog.dismiss()
                onDeleteListener.onDelete(user.userId)
            }

            binding.ivDeleteUser.setOnClickListener {
                dialogBinding.tvExplain.text = context.getString(
                    R.string.textview_text_are_you_sure_you_want_to_delete,
                    user.name
                )
                dialog.show()
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<UserModel>() {
            override fun areItemsTheSame(oldItem: UserModel, newItem: UserModel): Boolean =
                oldItem.userId == newItem.userId

            override fun areContentsTheSame(oldItem: UserModel, newItem: UserModel): Boolean =
                oldItem.name == newItem.name &&
                        oldItem.age == newItem.age &&
                        oldItem.countNumber == newItem.countNumber
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
        ViewHolder(ItemRvUserBinding.inflate(LayoutInflater.from(parent.context), parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        holder.bind(getItem(position))
}
